### PR TITLE
Add gateway to OvercloudNet

### PR DESCRIPTION
**Pull request by dprince**
The network info is now stored on each OvercloudIpset upon
creation.

Also, Updates baremetalset and controllervm to use the now
configurable but optional gateway on the ctlplane network.

### CODE REVIEW

File: controllers/baremetalset_controller.go [hunk 1]
```diff
@@ -477,7 +477,7 @@ func (r *BaremetalSetReconciler) baremetalHostProvision(instance *ospdirectorv1b
 	templateParameters = make(map[string]interface{})
 	templateParameters["MgmtIp"] = ip.String()
 	templateParameters["MgmtInterface"] = instance.Spec.MgmtInterface
-	templateParameters["MgmtGateway"] = "192.168.25.1" //FIXME
+	templateParameters["MgmtGateway"] = ipset.Status.Networks["ctlplane"].Gateway
 	templateParameters["MgmtNetmask"] = fmt.Sprintf("%d.%d.%d.%d", netMask[0], netMask[1], netMask[2], netMask[3])
 	// TODO: Either generate the gateway from the network object or allow user specification via CR?
```
Comment (abays) [inline controllers/baremetalset_controller.go:482]: We can remove this TODO now

File: controllers/overcloudipset_controller.go [hunk 2]
```diff
@@ -106,12 +107,13 @@ func (r *OvercloudIPSetReconciler) Reconcile(req ctrl.Request) (ctrl.Result, err
 			if err != nil {
 				if k8s_errors.IsNotFound(err) {
 					r.Log.Info(fmt.Sprintf("OvercloudNet named %s not found!", netName))
-					return ctrl.Result{}, nil
 				}
 				// Error reading the object - requeue the request.
 				return ctrl.Result{}, err
 			}
 
+			instance.Status.Networks[network.Name] = network.Spec
```
Comment (abays) [inline controllers/overcloudipset_controller.go:115]: Why do we store an entire `OvercloudNet` spec here instead of just its name?  Just curious.

Comment (dprince) [reply controllers/overcloudipset_controller.go:115]: So the 'Spec' has the information we need in it, namely the gateway. I thought about creating a duplicate struct for this but the struct has what we need and as this is just a cache it seemed reasonable.

File: controllers/baremetalset_controller.go (hunk 1 above)
Comment (dprince) [reply controllers/baremetalset_controller.go:482]: Ack. will do

Comment (dprince) [reply controllers/baremetalset_controller.go]: Fixed in the latest push.

File: templates/controllervm/cloudinit/networkdata [hunk 3]
```diff
@@ -1,4 +1,7 @@
 version: 2
 ethernets:
   eth1:
-    addresses: [ {{ .ControllerIP }} ]
\ No newline at end of file
+    addresses: [ {{ .ControllerIP }} ]
+{{if .Gateway }}
+    {{ .Gateway }}
```
Comment (stuggi) [inline templates/controllervm/cloudinit/networkdata:6]: this should be either gateway4: {{ .Gateway }} or gateway6: {{ .Gateway }}

https://cloudinit.readthedocs.io/en/latest/topics/network-config-format-v2.html#examples

Comment (stuggi) [reply templates/controllervm/cloudinit/networkdata:6]: oh wait ignore that, you do that when setting templateParameters["Gateway"]

File: controllers/baremetalset_controller.go [hunk 4]
```diff
@@ -477,9 +477,8 @@ func (r *BaremetalSetReconciler) baremetalHostProvision(instance *ospdirectorv1b
 	templateParameters = make(map[string]interface{})
 	templateParameters["MgmtIp"] = ip.String()
 	templateParameters["MgmtInterface"] = instance.Spec.MgmtInterface
-	templateParameters["MgmtGateway"] = "192.168.25.1" //FIXME
+	templateParameters["MgmtGateway"] = ipset.Status.Networks["ctlplane"].Gateway
```
Comment (stuggi) [inline controllers/baremetalset_controller.go:480]: right now this is not really Mgmt, this is Ctl, should we rename all of this and follow up with adding Mgmt net implementation?

File: templates/controllervm/cloudinit/networkdata (hunk 3 above)
Comment (dprince) [reply templates/controllervm/cloudinit/networkdata:6]: Yeah, I could have implemented it differently, this way seemed cleaner.

File: controllers/baremetalset_controller.go (hunk 4 above)
Comment (dprince) [reply controllers/baremetalset_controller.go:480]: Yes. We had agreed to do this just got forgotted I think. Anyways I filed this: https://github.com/openstack-k8s-operators/osp-director-operator/issues/92

**Review by stuggi — APPROVED** [review]

**Pull request merged**